File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int maxDepth = 0;
    boolean[] visited;

    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(k, dungeons, 0);
        return maxDepth;
    }

    private void dfs(int currentK, int[][] dungeons, int count) {
        maxDepth = Math.max(maxDepth, count);

        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && currentK >= dungeons[i][0]) {
                visited[i] = true;
                dfs(currentK - dungeons[i][1], dungeons, count + 1);
                visited[i] = false;
            }
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int max_count = 0; 
bool visited[8];

void dfs(int current_k, int count, const vector<vector<int>>& dungeons) {
    max_count = max(max_count, count);

    for (int i = 0; i < dungeons.size(); i++) {
        if (!visited[i] && current_k >= dungeons[i][0]) {
            visited[i] = true;
            
            dfs(current_k - dungeons[i][1], count + 1, dungeons);
            
            visited[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    max_count = 0;
    
    dfs(k, 0, dungeons);

    return max_count;
}